### PR TITLE
Prompt candidates with GCSE fail grades about how they are going to achieve passes

### PR DESCRIPTION
**Pull request by stevehook**
## Context

As a candidate who does not have a pass grade for English or maths (or science) GCSE I need to say how I plan to meet the requirement so that I can improve my chances of having a successful application.

## Changes proposed in this pull request

- [x] If a candidate does not enter a pass grade, we show an additional page in the flow: ‘You need a [subject] GCSE grade 4 (C) or above, or equivalent’, which gives candidates an opportunity to say how the expect to meet this requirement.
image

- [x] Show the response to the above question on the review page if applicable
image

- [x] Only for Maths, English and Science
- [x] Fix tests
- [x] Handle multi-grade awards

## Guidance to review

- I've not added a feature flag for this change. Is one needed?
- Can you see any issues with reusing the existing `ApplicationQualification#missing_explanation` attribute to store the new value? (Up to now it's only been used in cases where there is no qualification or grade).
- I've tried to handle single/multiple grade qualifications by taking the `grade` or extracting the individual grades from `constituent_grades`. Does this make sense? A pass is defined as a qualification with at least one pass grade.

## Link to Trello card

https://trello.com/c/ONl9m9RD/3307-ask-additional-question-if-candidate-does-not-enter-a-gcse-pass-grade

## Things to check

- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] This code does not rely on the addition/removal of Azure config environment variables in the same Pull Request
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/docs/environment-variables.md#azure-hosting-devops-pipeline)

### CODE REVIEW

File: app/models/application_qualification.rb [hunk 1]
```diff
@@ -144,6 +154,29 @@ def composite_equivalency_details
     details.strip if details.present?
   end
 
+  GCSE_PASS_GRADES = %w[A* A B C A*A* A*A AA AB BB BC CC CD 9 8 7 6 5 4 99 98 88 87 77 76 66 65 55 54 44 43].freeze
```
Comment (stevehook) [inline app/models/application_qualification.rb:157]: Taken directly from https://trello.com/c/ONl9m9RD/3307-ask-additional-question-if-candidate-does-not-enter-a-gcse-pass-grade

File: app/components/candidate_interface/gcse_qualification_review_component.rb [hunk 2]
```diff
@@ -75,6 +76,17 @@ def grade_row
       }
     end
 
+    def grade_explanation_row
```
Comment (davidgisbey) [inline app/components/candidate_interface/gcse_qualification_review_component.rb:79]: super nitty but maybe `failing_grade_explanation_row`

File: app/forms/candidate_interface/gcse_grade_explanation_form.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
+module CandidateInterface
+  class GcseGradeExplanationForm
+    include ActiveModel::Model
+
+    attr_accessor :missing_explanation
```
[the remark this answers is not in the document]
Comment (davidgisbey) [reply app/forms/candidate_interface/gcse_grade_explanation_form.rb:5]: At the moment if you select you are missing the qual then add a gcse with a failing grade, it prepopulates the missing box on the failing grade explanation page
image

[the remark this answers is not in the document]
Comment (davidgisbey) [reply app/forms/candidate_interface/gcse_grade_explanation_form.rb:5]: I actually think this is fine as we specify C or above in the title, so storing their response on a shared attr seems totally reasonable.

File: app/models/application_qualification.rb [hunk 4]
```diff
@@ -144,6 +154,29 @@ def composite_equivalency_details
...
+  def failed_required_gcse?
+    return true if required_gcse? &&
+      all_grades.present? &&
```
Comment (davidgisbey) [inline app/models/application_qualification.rb:160]: Super nitty, but i don't think we need to check for presence here

**Review by davidgisbey — APPROVED** [review]
Looks really great. Gave it a thorough test and it's working nicely. One super super nitty comment.

File: app/models/application_qualification.rb (hunk 4 above)
Comment (stevehook) [reply app/models/application_qualification.rb:160]: I had a test failure without this, can't remember the test, but it came down to a qualification that didn't (yet) have a grade at all being flagged as failed. That doesn't make a lot of sense - you can't be flagged as failed if you've not entered a grade yet.

**Pull request merged**